primary_beam_correction.py

```diff
@@ -0,0 +1,296 @@
+    '''
+    images = read_fits(path)
+    c_freq_plane = []
+    for i in range(1, images.header['NSPEC']+1):
+        c_freq_plane.append(images.header['FREQ{0:04}'.format(i)])
+    return np.array(c_freq_plane)
+
+
+def beam_pattern(separation_rad, path, c_freq=None):
+    '''
+    Making beam pattern using the DEEP2 paper
+    Input : Radial seperation array and path to the fits file
+    Output : Beam pattern
+    '''
+    if c_freq:
+        rho = separation_rad
+        c = c_freq*1e6
+        # Degrees to radians conversion
+        conv = np.pi / 180.
+        # converting arcminutes to radians
+        v_beam_rad = 89.5 / 60. * conv
+        h_beam_rad = 86.2 / 60. * conv
+        # Taking the Geometric mean for the veritical and horizontal cut through the beam.
+        vh_beam_mean = np.sqrt(v_beam_rad*h_beam_rad)
+        flux_density = []
+        nu = c/1.e9
+        theta_b = vh_beam_mean*(1./nu)
+        ratio = rho/theta_b
+        num = np.cos(1.189*np.pi*ratio)
+        dem = 1-4*(1.189*ratio)**2
+        a_b = (num/dem)**2
+        flux_density.append(a_b)
+    else:
+        rho = separation_rad
+        c = central_freq(path)
+        # Degrees to radians conversion
+        conv = np.pi / 180.
+        # converting arcminutes to radians
+        v_beam_rad = 89.5 / 60. * conv
+        h_beam_rad = 86.2 / 60. * conv
+        # Taking the Geometric mean for the veritical and horizontal cut through the beam.
+        vh_beam_mean = np.sqrt(v_beam_rad*h_beam_rad)
+        flux_density = []
+        for nu in c:
+            nu = nu/1.e9
+            theta_b = vh_beam_mean*(1./nu)
+            ratio = rho/theta_b
+            num = np.cos(1.189*np.pi*ratio)
+            dem = 1-4*(1.189*ratio)**2
+            a_b = (num/dem)**2
+            flux_density.append(a_b)
+    return flux_density
+
+
+def mad(data):
+    '''
+    Calculating meadian absolute deviation (MAD).
+    '''
+    MAD_TO_SD = 1.4826
+    med = np.nanmedian(data)
+    dev = np.abs(data - med)
+    return med, MAD_TO_SD * np.nanmedian(dev)
+
+
+def standard_deviation(data):
+    '''
+    Calcalating the standard deviation of each frequency plane using MAD.
+    Returns: Weights per each frequency plane
+    '''
+    diff = 1
+    i = 0
+    data = data[data != 0.0]
+    if len(data) == 0:
+        return float(0.0), float(0.0)
+    med, sd = mad(data)
+    while diff > 0:
+        i += 1
+        if i > 50:
+            return sd
+        old_sd = sd
+        old_len = len(data)
+        cut = np.abs(data - med) < 5.0 * sd
+        if np.all(~cut):
+            return sd
+        data = data[cut]
+        sd = mad(data)[1]
+        diff = old_len - len(data)
+        if sd == 0.0:
+            return old_sd
+    return 1/(sd)**2
+
+
+def weigted_average(arr, weights):
+    '''
+    Computing weighted average of all the frequency planes.
+    '''
+    wt_average = np.average(arr, weights=weights, axis=0)
+    return wt_average
```

Comment:
This will work to blank pixels outside the 10% level of the beam at the largets frequency but it will return nans everywhere if just one of the planes in the input image is blanked. I haven't noticed anything anywhere else in the code that deals with this case - is it there?

Reply:
No there isn't, but initially, my function was dealing with it but you advised that I remove it. You said if on one of the planes we find a nan value on a certain frequency then we return a nan. 

```
Below it is the function that I was using before:
```def weigted_average(arr, weights):
    '''
    Computing weighted average of all the frequency planes.
    '''
    wt_average = []
    for i in range(arr.shape[1]):
        indices = ~np.isnan(arr[:,i])
        w = weights[indices]
        a = arr[:,i][indices]
        if len(a) == 0:
            wt_average.append(np.nan)
        else:
            wt_average.append(np.average(a, weights = w))
    return np.array(wt_average)